### DISCUSSION TITLE
Trouble with boolean operations

### DISCUSSION THREAD

**Discussion by JohannesMerkt** [donalffons/opencascade.js#143] [Q&A] [answered]
I have some weird issues with boolean operations. It seems like i have set up some resolution settings incorrectly. Here is my difference operation code:

```ts
    if (shape1 && shape2) {
        const range1 = new openCascade.Message_ProgressRange_1(); // why is this needed?
        const differenceCut = new openCascade.BRepAlgoAPI_Cut_3(shape1, shape2, range1);
        differenceCut.SetFuzzyValue(0.1); // what does this value do?
        const range2 = new openCascade.Message_ProgressRange_1(); // another range?
        differenceCut.Build(range2);
        resolve([differenceCut.Shape()]);
    }
```
Ive let typescript guide me. It told me i need to use ranges, I dont know however why i need them.

Here some images using that difference operation:
On the left the box i want to subtract from. In center the box im subtracting. Right the result
<span><img src="https://user-images.githubusercontent.com/25324640/171393853-55ef4d66-6fe9-49ba-a150-64a4d1346e99.png" width="200" /></span><span><img src="https://user-images.githubusercontent.com/25324640/171393915-d2fd13bc-713f-42f3-86b1-4bb14b7d4b6d.png" width="200" /></span><span><img src="https://user-images.githubusercontent.com/25324640/171393966-bcf3bbdf-d829-4502-aff5-7c1216fc1938.png" width="200" /></span>
But if I translate the box inside on the red x axis so that one side gets a thinner wall. Things dont work anymore:
<img src="https://user-images.githubusercontent.com/25324640/171394018-d096f94f-06bd-42fa-85ac-d660f81a1192.png" width="200" />
Similarly if the box inside gets offset on the blue z axis i suddently get sloped sides which shouldnt happen
<span><img src="https://user-images.githubusercontent.com/25324640/171396815-3fb72e9e-d582-4fc5-9045-5e59531510de.png" width="200" /></span><span><img src="https://user-images.githubusercontent.com/25324640/171394172-58e1b295-de34-4b11-a3ea-6df422134bbc.png" width="200" /></span>

I have similar problems with Common and Fuse. Is there a way to make these more reliable?

**Comment by donalffons** [comment, marked as answer]
Couple of notes:
* The `new oc.Message_ProgressRange_1()` is used by OpenCascade to coordinate progress indicators for one or more long-running tasks. In C++, those arguments contain default values, but since OpenCascade.js currently doesn't support default values, you have to specify them explicitly. There are a few more notes on the topic of progress indicators [in the docs](https://ocjs.org/docs/advanced/progress-indicators-user-break/intro) and in the OCCT forums.
* I can't give you an exact definition of what `SetFuzzyValue` does, only so much: It is some sort of tolerance value that is used when performing boolean operations. Geometry might get collapsed if distances are below that value. The reason for your problems is almost certainly related to the `FuzzyValue` being too large. I think that value should be significantly smaller than the minimum length of any geometry that you want to keep in your model.
* I believe you don't need the call to `differenceCut.Build()`, since the [constructor you're using is already making that call for you](https://github.dev/Open-Cascade-SAS/OCCT/blob/9ff36dac47b6f3bac4a433e14832d2937a0a7d00/src/BRepAlgoAPI/BRepAlgoAPI_Cut.cxx#L59).

> I have similar problems with Common and Fuse. Is there a way to make these more reliable?

I would suggest to use a much smaller `FuzzyValue`. If your boolean operations are between relatively simple shapes you should be able to get them stable. If you're operating with more complex shapes, my experience is that you should be prepared for them to fail sometimes.

**Reply by JohannesMerkt** [reply]
Thank you very much for this detailed answer! I tried decreasing the fuzzyValue and it helped immediatly. Also the differenceCut.Build() was not needed.

**Comment by sridhar-mani** [comment]
is this a wasm build function or directly from the available opencascade.js
